syntax = "proto3";

message Void {}

service Metrics {
  rpc GetMetrics(MetricsRequest) returns (MetricsResponse);
}

message Metric {
  string name = 1;
  float value = 2;
}

message MetricsResponse {
  string name = 1;
  repeated Metric metrics = 2;
}

message MetricsRequest {
  string base_date = 1;
  string ticker = 2;
}


service Info {
    rpc GetInfo(InfoRequest) returns (InfoResponse);
}

message Value {
  string date = 1;
  float value = 2;
}

message InfoRequest {
  string ticker = 1;
  string start_date = 2;
  string end_date = 3;
}

message InfoResponse {
  string name = 1;
  repeated Value prices = 2;
  repeated Value volumes = 3;
  repeated Value dividends = 4;
}



service Products {
  rpc GetProducts(Void) returns (stream LoadedProducts);
}

message Notification {
  string email = 1;
  LoadedProducts products = 2;
}

message StockFII {
  string name = 1;
  string ticker = 3;
  string category = 2;
  string link = 4;
  string segment = 5;
  string recommendation = 6;
  string target_price = 7;
  string publication_date = 8;
  string description = 9;
  float price = 10;
  float potential = 11;
  float return12m = 12;
  float forecast12m = 13;
  float mean_daily_volume = 14;
  float mean_daily_return = 15;
  float dividend_yield = 16;
}

message COE {
  string name = 1;
  string strategy = 2;
  string classification = 3;
  int32 deadline = 4;
  string description = 5;
  string category = 6;
  string link = 7;
}

message FixedIncome {

}

message InvestmentFund {
  string name = 1;
  string link = 2;
  string category = 3;
  float min_application = 4;
  float adm_tax = 5;
  string redemption = 6;
  string classification = 7;
  float month_return = 8;
}

message PensionFund {
  string name = 1;
  float min_application = 2;
  float month_return = 3;
  float year_return = 4;
  float adm_tax = 5;
  string category = 6;
}

message LoadedProducts {
  string date = 1;
  repeated StockFII stocks = 2;
  repeated StockFII real_state = 3;
  repeated COE coe = 4;
  repeated FixedIncome fixed_income = 5;
  repeated InvestmentFund investment_funds = 6;
  repeated PensionFund pension_funds = 7;
}


